File: archimedes-singularity/archimedes-r/functions/dataflow/R/path.r
input_path <- function(
  key # a string or index
) {

  if (is.character(key) && ! key %in% names(snakemake@input)) {
    stop("No input keyed ", key, " exists in the snakemake object for this job.")
  }
  
  path <- snakemake@input[[key]]

  if (!file.exists(path)) {
    stop("The input file for key ", key, " does not exist for this job.")
  }

  path
}

input_single_var <- function(
  key,
  type
) {
  scan(
    input_path(key),
    what = type,
    quiet = TRUE)
}

input_str <- function(key) {
  input_single_var(key, character())
}

input_int <- function(key) {
  input_single_var(key, integer())
}

input_num <- function(key) {
  input_single_var(key, numeric())
}

input_bool <- function(key) {
  # Checks is input, as string, is a TRUE-like string.
  # Else FALSE.
  str <- input_single_var(key, character())
  tolower(str) %in% c("yes", "y", "true", "t", "1")
}

input_json <- function(key) {
  jsonlite::fromJSON(input_path(key), simplifyMatrix = FALSE)
}

output_path <- function(
  key # a string or index
) {

  if (is.character(key) && ! key %in% names(snakemake@output)) {
    stop("No output keyed ", key, " exists in the snakemake object for this job.")
  }
  
  snakemake@output[[key]]
}

output_string <- function(
  data,
  key
) {
  if (is.character(data)) {
    data <- paste0("\"",data,"\"")
  }

  write(data, output_path(key))
}

output_var <- function(
  data,
  key
) {

  write(data, output_path(key))
}

output_json <- function(data, key) {
  json_str <- jsonlite::toJSON(
    data,
    null = "null",
    na = "string",
    factor = "string"
  )
  write(json_str, output_path(key))
}

File: archimedes-singularity/archimedes-py/functions/dataflow/path.py
# Do not let these imports become usable within scripts themselves
import os as _os
import os.path as _os_path
from pandas import DataFrame, read_csv
import json
import tarfile

def input_path(key, snakemake):
    if isinstance(key, str) and not key in snakemake.input:
        raise ValueError(f"The input key {key} does not exist for this job.")
    path = snakemake.input[key]
    if not _os_path.exists(path):
        raise ValueError(f"The input file for key {key} does not exist for this job.")
    return path

def input_var(key, snakemake):
    return open(input_path(key, snakemake), 'r').read()

def input_tsv(key, snakemake):
    return read_csv(input_path(key, snakemake) , sep='\t')

def input_json(key, snakemake):
    return json.load(open(input_path(key, snakemake)))

def input_bool(key, snakemake):
    def str2bool(str):
        return str.lower() in ("yes", "y", "true", "t", "1")
    return  str2bool(input_var(key, snakemake))

def param(key, snakemake):
    if isinstance(key, str) and not key in snakemake.param:
        raise ValueError(f"The params key {key} does not exist for this job.")
    return snakemake@param[key]

def param_path(key, snakemake):
    path = param(key, snakemake)
    if not _os_path.exists(path):
        raise ValueError(f"No file at param key {key} path of {path}.")
    return path

def output_path(key, snakemake):
    if isinstance(key, str) and not key in snakemake.output:
        raise ValueError(f"The output key {key} does not exist for this job.")
    return snakemake.output[key]

def output_tsv(data, key, snakemake):
    return data.to_csv(output_path(key, snakemake) , sep='\t')

def output_json(data, key, snakemake):
    with open(output_path(key, snakemake), 'w') as output_file:
        output_file.write(json.dumps(data))

def output_var(data, key, snakemake):
    with open(output_path(key, snakemake), 'w') as output_file:
        output_file.write(data)

def output_tgz(source_dir, key, snakemake):
    with tarfile.open(output_path(key, snakemake), "w:gz") as tar:
        tar.add(source_dir, arcname=_os_path.basename(source_dir))
